### PR TITLE
Support a wider gamut of `rand` operations

### LINKED ISSUE #63: `rand` should support types / sets of values
Currently `rand` supports generating floats, and arrays thereof, but you can't specify e.g. integers or other sets of values from which random numbers should be drawn.

### PR DESCRIPTION
**Pull request by tpgillam**
Closes #63

### CODE REVIEW

**Comment by tpgillam** [comment]
@willtebbutt I'm actually wondering if it would be good to make a breaking change here, whilst working on this section.

Currently, this version of `rand` takes:
* A node `x`, which is used only for alignment
* (maybe) a constant `S`, determining the sampleable
* (maybe) some dimensions `d` (either tuple or int varargs), which make the output value type be an array

I think it might be more future proof to unify `x` and `S`.

*Pros*: it means that in the future, one can e.g. have a node that ticks with distributions, so that the random value drawn at each knot can be drawn from that distribution.

*Cons*: It is breaking (although better now than later), and it's slightly more awkward to generate "trivial" random numbers.

e.g. old:
```julia
x = pulse(Hour(1))
y = rand(x)
```

new:
```
S = align(Float64, pulse(Hour(1)))  # i.e. a node that ticks with `Float64` every time that we wish to sample
y = rand()
```

File: test/ops/random.jl [hunk 1]
```diff
@@ -24,25 +25,53 @@
...
+        @test n != rand(x, Int32)
+        @test value_type(n) == Int32
+
+        @testset "collections of values" begin
+            for vals in (
+                (1, 2.0, 2.0 + 0im),
+                [1, 2],
+                [1 2; 3 4],
+                Set([1, 2, 3]),
+                Dict(1 => "1", 2 => "2", 3 => "3"),
```
Comment (tpgillam) [inline test/ops/random.jl:40]: Refactor out (repeated below).
Add a range.

File: test/ops/random.jl [hunk 2]
```diff
@@ -24,25 +25,53 @@
...
+            )
+                n = rand(rng, x, vals)
+                @test n === rand(rng, x, vals)
+                @test value_type(n) <: foldl(Base.promote_typejoin, typeof.(collect(vals)))
```
Comment (tpgillam) [inline test/ops/random.jl:44]: Check that `value_type(n)` is a SUPERTYPE of every element in vals.
Same below in array case.

**Review by willtebbutt — APPROVED** [review]
LGTM

**Pull request merged**